Populate Google tag data for existing sites

nfmohit opened the pull request:
## Summary

<!-- Please reference the issue this PR addresses in the following list. -->
Addresses issue:

- #6082 

## Relevant technical choices

This PR makes Site Kit populate Google Tag data for existing sites using GA4.

## PR Author Checklist

- [ ] My code is tested and passes existing unit tests.
- [ ] My code has an appropriate set of unit tests which all pass.
- [ ] My code is backward-compatible with WordPress 5.2 and PHP 5.6.
- [ ] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [ ] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [x] Run the code.
- [x] Ensure the acceptance criteria are satisfied.
- [x] Reassess the implementation with the IB.
- [x] Ensure no unrelated changes are included.
- [x] Ensure CI checks pass.
- [ ] Check Storybook where applicable.
- [x] Ensure there is a QA Brief.

## Merge Reviewer Checklist

- [ ] Ensure the PR has the correct target branch.
- [ ] Double-check that the PR is okay to be merged.
- [ ] Ensure the corresponding issue has a ZenHub release assigned.
- [ ] Add a changelog message to the issue.

techanvil commented on assets/js/modules/analytics-4/datastore/properties.test.js:
```diff
@@ -607,6 +611,180 @@ describe( 'modules/analytics-4 properties', () => {
+
+				expect(
+					registry.select( MODULES_ANALYTICS_4 ).getGoogleTagID()
+				).toEqual( '' );
```
We may as well add another assertion to check the value of `googleTagLastSyncedAtMs` is unchanged here, to help make it clear the sync has not been attempted...

nfmohit replied:
Added, thank you!

techanvil commented on assets/js/modules/analytics-4/datastore/properties.test.js:
```diff
@@ -607,6 +611,180 @@ describe( 'modules/analytics-4 properties', () => {
+				} );
+
+				fetchMock.getOnce( googleTagSettingsEndpoint, {
+					body: fixtures.googleTagSettings,
```
It would be worth passing a deep copy of `fixtures.googleTagSettings` here, so we can be sure that the code being executed for the test does not inadvertently change the values we're using in the assertions.

```suggestion
					body: cloneDeep( fixtures.googleTagSettings ),
```

nfmohit replied:
Updated, thank you!

techanvil commented on assets/js/modules/analytics-4/datastore/properties.test.js:
```diff
@@ -607,6 +611,180 @@ describe( 'modules/analytics-4 properties', () => {
+					.syncGoogleTagSettings();
+
+				expect( fetchMock ).toHaveFetchedTimes( 2 );
+				expect( fetchMock ).toHaveFetched( googleTagSettingsEndpoint, {
```
As we're asserting the `GET` to `googleTagSettingsEndpoint`, we should do the same for the `POST` to `ga4SettingsEndpoint`...

nfmohit replied:
Updated, thank you!

techanvil commented on assets/js/util/date-range/constants.js:
```diff
@@ -26,3 +26,4 @@ export const INVALID_DATE_RANGE_ERROR =
 export const HOUR_IN_SECONDS = 3600;
 export const DAY_IN_SECONDS = 86400;
 export const WEEK_IN_SECONDS = 604800;
+export const HOUR_IN_MILLISECONDS = HOUR_IN_SECONDS * 1000;
```
As mentioned, I think we should probably _not_ add this constant at this point, but consider a followup issue to do so across the codebase...

nfmohit replied:
Removed, thank you!

techanvil commented on assets/js/modules/analytics-4/datastore/properties.js:
```diff
@@ -490,6 +498,53 @@ const baseActions = {
+			__experimentalResolveSelect( CORE_MODULES ).getModules()
+		);
+
+		const { isModuleConnected } = select( CORE_MODULES );
```
Is there a functional reason this line is between the two `yield` statements?

If not, I'd suggest moving it after them, and closer to where the selector is used, for a marginal improvement in organisation.

nfmohit replied:
I agree. However, I have re-structured the selector as a part of [this commit](https://github.com/google/site-kit-wp/pull/6536/commits/5d61a1f6ddbf2b58c033183dd5741813522aa964) to return early to prevent continuing with selections unnecessarily. Does this look good to you?

techanvil replied:
Thanks, yes that looks much better, nice one :)

techanvil reviewed (changes requested):
Nice work here @nfmohit. I've left a few comments, but all very minor points. Please take a look.

techanvil commented on assets/js/modules/analytics-4/datastore/properties.js:
```diff
@@ -451,9 +453,15 @@ const baseActions = {
+			.select( MODULES_ANALYTICS_4 )
+			.getMeasurementID();
+
+		if ( measurementID !== existingMeasurementID ) {
```
Would you mind explaining what this change was needed for?

nfmohit replied:
Sure!

In this PR, we have used the `updateSettingsForMeasurementID()` action inside the `syncGoogleTagSettings()` action. In this particular case, the primary aim was not to update the measurement ID, but only the relevant settings (i.e. Google Tag settings).

For such cases, I think dispatching the `setMeasurementID()` action and thus resetting the measurement ID is unnecessary. Hence, I've set it to only run if the passed measurement ID is actually different from the current measurement ID.

Do you think this is valid?

I have also added a comment, however, if you think otherwise, I'll revert this change entirely.

Thanks!

techanvil replied:
Thanks for clarifying. However, I'm still a little confused - you've said this is to avoid resetting the measurement ID, but if `measurementID === existingMeasurementID` (i.e. the case you're now avoiding), and `setMeasurementID( measurementID )` is invoked, it's just the same measurement ID anyway so should not have any practical effect. The term "reset" is a bit unclear as it makes it sound as though the setting would be changing, whereas it would simply remain the same.

That's why I've a bit unclear as to why this is needed. Unless there's a clear side effect of calling `setMeasurementID()` with the same value, I don't think we _need_ to add this condition - although, there's also nothing _wrong_ with it either, it was just a bit confusing to see due to the lack of clarity about why it's added in this PR.

Regarding unwanted side effects, I don't think there should be one, as `haveSettingsChanged()` / `hasSettingChanged()` do an equality check on the setting/saved setting, but maybe you've spotted something else. However if not, I think it should be safe to remove this condition.

nfmohit replied:
You're right. I think I was being too cautious here. Removed, thanks!

nfmohit commented on assets/js/modules/analytics-4/datastore/properties.js:
```diff
@@ -490,6 +498,53 @@ const baseActions = {
+		if (
+			isFeatureEnabled( 'gteSupport' ) &&
+			isGA4Connected &&
+			isEmpty( googleTagID ) &&
+			! isEmpty( measurementID ) &&
```
Thank you for flagging this, @techanvil!

I am working on removing the unnecessary `isEmpty()` here. However, in some cases, for example, setting up Analytics without the `gteSupport` feature flag, the `googleTagID` property does exist in the GA4 settings, but as an empty string. Screenshot:

![image](https://user-images.githubusercontent.com/20284937/219445114-4ac626c1-e3ea-4faf-9e71-9772fec65f8d.png)

In such a case, do you think it'd be wise to keep using `isEmpty()` for `googleTagID`?

Thanks!

nfmohit replied:
Removed `isEmpty()` for `measurementID` for now, thank you!

techanvil replied:
Thanks @nfmohit. Regarding `googleTagID`, it's a string, so I don't see any need to use `isEmpty()` to check its value. Unless I am missing something, truthy/falsy should be perfectly OK here...

nfmohit replied:
Thank you for the clarification, @techanvil! I've removed it.

nfmohit commented on assets/js/components/DashboardEntryPoint.test.js:
```diff
@@ -19,24 +19,32 @@
+		provideModules( registry );
+		registry.dispatch( MODULES_ANALYTICS_4 ).receiveGetSettings( {} );
+	};
 	it( 'should render the unified dashboard', () => {
```
Updated, thank you!

nfmohit commented on assets/js/modules/analytics-4/datastore/properties.js:
```diff
@@ -490,6 +498,53 @@ const baseActions = {
+			isEmpty( googleTagID ) &&
+			! isEmpty( measurementID ) &&
+			( ! googleTagLastSyncedAtMs ||
+				Date.now() - googleTagLastSyncedAtMs >= HOUR_IN_MILLISECONDS )
```
Updated, will open an issue about this, thanks!

nfmohit commented on assets/js/modules/analytics-4/datastore/properties.test.js:
```diff
@@ -607,6 +611,180 @@ describe( 'modules/analytics-4 properties', () => {
+				registry.dispatch( MODULES_ANALYTICS_4 ).receiveGetSettings( {
+					measurementID: 'G-1A2BCD346E',
+					googleTagID: 'GT-12345',
+					googleTagLastSyncedAtMs: 1676519625246,
```
Updated, thank you!

nfmohit commented:
> Nice work here @nfmohit. I've left a few comments, but all minor points. Please take a look.

Thank you so much for the detailed and kind review, @techanvil! I have addressed all the comments, and left some follow-up questions on a few. Thank you!

techanvil commented on assets/js/modules/analytics-4/datastore/properties.js:
```diff
@@ -490,6 +501,58 @@ const baseActions = {
+		const googleTagLastSyncedAtMs = getGoogleTagLastSyncedAtMs();
+
+		if (
+			isEmpty( googleTagID ) &&
```
As mentioned in my reply to your question, I still don't see any reason to use `isEmpty()` here over a regular `! googleTagID`... Calling `isEmpty()` should have the same result but there's no need to use a third-party utility...

techanvil reviewed (changes requested):
Thanks for making these changes @nfmohit. I've left a couple of replies to your comments. Once these are resolved, this looks good to go.

techanvil reviewed (approved):
LGTM, nice one @nfmohit!
